### PR TITLE
Refactor Acquire Token APIs Internally

### PR DESCRIPTION
**Pull request by pkanher617**
This PR is to refactor our code for interactive acquire token calls, which will avoid repetition. There are some considerations to make here, including:

- allowing request.account in login calls
- removal of silentLogin
- remove requirement to call login before acquireToken
- acquireToken and loginInProgress call changes

The above changes can come in separate PRs that will be released in a major version, since these will be breaking changes.

Here is a summary of changes in this PR:
- Refactored all interactive calls to a single function
- fixed bug when sending clientId as only scope (?)
- Renamed isCallback function to isStsResponse
- Added new success and error handlers which will handle all interactive errors (will remove the old success handlers for redirect)
- promptUser function renamed to navigateWindow, now takes a popup window, will navigate popups when given a window object
- Moved around functions

### CODE REVIEW

File: lib/msal-core/src/Constants.ts [hunk 1]
```diff
@@ -80,6 +80,9 @@ export class Constants {
 
   static get cacheLocationLocal(): CacheLocation { return "localStorage"; }
   static get cacheLocationSession(): CacheLocation { return "sessionStorage"; }
+
+  static get interactionTypeRedirect(): InteractionType { return "redirectInteraction"; }
+  static get interactionTypePopup(): InteractionType { return "popupInteraction"; }
```
Comment (DarylThayil) [inline lib/msal-core/src/Constants.ts]: are there other places this could go? This file is already so bloated

**Comment by DarylThayil** [comment]
what are the breaking changes?

**Comment by DarylThayil** [comment]
Might be good to schedule some time to go over this as a team, a lot of code changed and hard to ngrok it all

File: lib/msal-core/src/Constants.ts (hunk 1 above)
Comment (pkanher617) [reply lib/msal-core/src/Constants.ts]: I tried to keep it organized, we should probably have a separate PR to refactor the Constants file. I am open to suggestions.

**Comment by pkanher617** [comment]
> what are the breaking changes?

There aren't any breaking changes yet, I should have edited the description. The considerations I have listed would be breaking changes depending on what we decide.

> Might be good to schedule some time to go over this as a team, a lot of code changed and hard to ngrok it all

I will throw something on the calendar.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 2]
```diff
@@ -1276,6 +1077,24 @@ export class UserAgentApplication {
...
 
+  /**
+   * @hidden
+   * @ignore
+   * Checks if the redirect response is received from the STS. In case of redirect, the url fragment has either id_token, access_token or error.
+   * @param {string} hash - Hash passed from redirect page.
+   * @returns {Boolean} - true if response contains id_token, access_token or error, false otherwise.
+   */
+  urlContainsHash(hash: string): boolean {
+    hash = this.getHash(hash);
```
Comment (jasonnutter) [inline lib/msal-core/src/UserAgentApplication.ts]: I think the parameter should be called `url`, instead of hash:

```js
urlContainsHash(url: string): boolean {
    const hash = this.getHash(url );
```

Comment (jasonnutter) [reply lib/msal-core/src/UserAgentApplication.ts]: Or if a hash value if the only value that should be passed to this, maybe change the name to `isStsResponse` or something?

Comment (sameerag) [reply lib/msal-core/src/UserAgentApplication.ts]: It is mostly called hash as we read only the fragment of the STSResponse to populate the hash.

**Comment by sameerag** [comment]
I am not adding comments as we will go through this tomorrow in the review meetup. We need to follow this up with:

1) The much-needed split of UserAgentApplication.ts file as sectioned already
2) Break large functions in UAA to smaller ones (saveHashToken, processCallBack etc ...)

I know you are working on some PRs regarding this, I can help too, let me know and we can split up.

File: lib/msal-core/src/UserAgentApplication.ts (hunk 2 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: Thanks for the feedback guys, I made some changes. I will push and we can review later.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 3]
```diff
@@ -1157,11 +953,12 @@ export class UserAgentApplication {
...
    * @param {string} urlNavigate - URL of the authorization endpoint
    */
-  private promptUser(urlNavigate: string) {
+  private promptUser(urlNavigate: string, popupWindow?: Window) {
     // Navigate if valid URL
     if (urlNavigate && !Utils.isEmpty(urlNavigate)) {
-      this.logger.infoPii("Navigate to:" + urlNavigate);
-      window.location.replace(urlNavigate);
+      let navigateWindow: Window = popupWindow ? popupWindow : window;
+      popupWindow ? this.logger.infoPii("Navigated Popup window to:" + urlNavigate) : this.logger.infoPii("Navigate to:" + urlNavigate);
```
[the remark this answers is not in the document]
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: Thanks, great feedback

File: lib/msal-core/src/UserAgentApplication.ts [hunk 4]
```diff
@@ -1109,6 +893,18 @@ export class UserAgentApplication {
...
+  /**
+   * @hidden
+   */
+  private isInteractionRequired(errorString: string) : boolean {
+    if (errorString.indexOf("interaction_required") !== -1 ||
+    errorString.indexOf("consent_required") !== -1 ||
+    errorString.indexOf("login_required") !== -1) {
+      return true;
+    }
+    return false;
```
[the remark this answers is not in the document]
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: Thanks, I made the change. I also moved this function into the InteractionRequiredAuthError class.

**Comment by pkanher617** [comment]
isCallback is a breaking change

File: lib/msal-core/src/UserAgentApplication.ts [hunk 5]
```diff
@@ -303,122 +303,40 @@ export class UserAgentApplication {
...
 
-  //#endregion
+  private successHandler(interactionType: InteractionType, response: AuthResponse, resolve?: any) : void {
+    if (interactionType === Constants.interactionTypeRedirect) {
+      if (this.errorReceivedCallback) {
+        this.tokenReceivedCallback(response);
+      } else if (this.authResponseCallback) {
+        this.authResponseCallback(null, response);
+      }
+    } else {
```
Comment (pkanher617) [inline lib/msal-core/src/UserAgentApplication.ts]: add else-if case for popup

Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: and error for else cases

File: lib/msal-core/src/UserAgentApplication.ts [hunk 6]
```diff
@@ -428,120 +346,77 @@ export class UserAgentApplication {
...
-      // Fail if login is already in progress
-      if (this.loginInProgress) {
-        return reject(ClientAuthError.createLoginInProgressError());
-      }
+  private acquireTokenInteractive(interactionType: InteractionType, isLoginCall: boolean, request?: AuthenticationParameters, resolve?: any, reject?: any): void {
+
+    // If already in progress, do not proceed
+    if (this.loginInProgress || this.acquireTokenInProgress) {
+      this.errorHandler(interactionType,
+        this.loginInProgress ? ClientAuthError.createLoginInProgressError() : ClientAuthError.createAcquireTokenInProgressError(),
```
Comment (pkanher617) [inline lib/msal-core/src/UserAgentApplication.ts]: Improved readability here.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 7]
```diff
@@ -428,120 +346,77 @@ export class UserAgentApplication {
...
+        buildResponseStateOnly(this.getAccountState(request && request.state)),
+        reject);
+      return;
+    }
 
+    let scopes: Array<string>;
```
Comment (pkanher617) [inline lib/msal-core/src/UserAgentApplication.ts]: Add ternary here.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 8]
```diff
@@ -550,200 +425,246 @@ export class UserAgentApplication {
...
-
-    // Track login progress
-    this.loginInProgress = true;
-
-    // Resolve endpoint
-    this.authorityInstance.resolveEndpointsAsync().then(() => {
-      let serverAuthenticationRequest = new ServerRequestParameters(this.authorityInstance, this.clientId, scopes, ResponseTypes.id_token, this.getRedirectUri(), request && request.state);
+  private acquireTokenHelper(account: Account, interactionType: InteractionType, isLoginCall: boolean, request?: AuthenticationParameters, scopes?: Array<string>, resolve?: any, reject?: any): void {
+    // Track the acquireToken progress
+    isLoginCall ? this.loginInProgress = true : this.acquireTokenInProgress = true;
```
Comment (pkanher617) [inline lib/msal-core/src/UserAgentApplication.ts]: if statement

File: lib/msal-core/src/UserAgentApplication.ts [hunk 9]
```diff
@@ -550,200 +425,246 @@ export class UserAgentApplication {
...
 
-      // populate QueryParameters (sid/login_hint/domain_hint) and any other extraQueryParameters set by the developer;
-      serverAuthenticationRequest = this.populateQueryParams(account, request, serverAuthenticationRequest);
+    const scope = scopes ? scopes.join(" ").toLowerCase() : this.clientId.toLowerCase();
```
Comment (pkanher617) [inline lib/msal-core/src/UserAgentApplication.ts]: Check if scopes util is created for this, otherwise add one.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 10]
```diff
@@ -550,200 +425,246 @@ export class UserAgentApplication {
...
+      if (interactionType === Constants.interactionTypeRedirect) {
+        if (!isLoginCall) {
+          this.cacheStorage.setItem(Constants.stateAcquireToken, serverAuthenticationRequest.state, this.inCookie);
+        }
+      } else {
+        window.renewStates.push(serverAuthenticationRequest.state);
+        if (isLoginCall) {
+          window.requestType = Constants.login;
+        } else {
+          window.requestType = Constants.renewToken;
```
Comment (jasonnutter) [inline lib/msal-core/src/UserAgentApplication.ts]: This can be a ternary.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 11]
```diff
@@ -550,200 +425,246 @@ export class UserAgentApplication {
...
+  acquireTokenSilent(request: AuthenticationParameters): Promise<AuthResponse> {
     return new Promise<AuthResponse>((resolve, reject) => {
+
       // Validate and filter scopes (the validate function will throw if validation fails)
       this.validateInputScope(request.scopes, true);
 
       const scope = request.scopes.join(" ").toLowerCase();
 
-      // Get the account object if a session exists
+      // if the developer passes an account give him the priority
```
Comment (jasonnutter) [inline lib/msal-core/src/UserAgentApplication.ts]: I'd reword this to something like "Passed in account should take precedence"

File: lib/msal-core/src/Utils.ts [hunk 12]
```diff
@@ -538,6 +538,24 @@ export class Utils {
...
+  static getHashFromUrl(urlStringOrFragment: string): string {
+    let hash = urlStringOrFragment;
+    const hashIndex1 = urlStringOrFragment.indexOf("#");
+    const hashIndex2 = urlStringOrFragment.indexOf("#/");
+    if (hashIndex2 > -1) {
+      hash = urlStringOrFragment.substring(hashIndex2 + 2);
+    } else if (hashIndex1 > -1) {
+      hash = urlStringOrFragment.substring(hashIndex1 + 1);
+    }
+    return hash;
```
Comment (jasonnutter) [inline lib/msal-core/src/Utils.ts]: Instead of reassigning, you can early return:

```js
if (hashIndex2 > -1) {
  return urlStringOrFragment.substring(hashIndex2 + 2);
}

if (hashIndex1 > -1) {
   return urlStringOrFragment.substring(hashIndex1 + 1);
}

return hash;
```

File: lib/msal-core/src/UserAgentApplication.ts [hunk 13]
```diff
@@ -1415,7 +1222,7 @@ export class UserAgentApplication {
    * @param hash
    */
   private deserializeHash(hash: string) {
-    hash = this.getHash(hash);
+    hash = Utils.getHashFromUrl(hash);
```
Comment (jasonnutter) [inline lib/msal-core/src/UserAgentApplication.ts]: Reassigning parameter values can be dangerous, I would like to see this as a new variable (or even just passed in directly into `Utils.deserialie`.

File: lib/msal-core/src/UserAgentApplication.ts [hunk 14]
```diff
@@ -550,200 +425,246 @@ export class UserAgentApplication {
...
+        this.logger.infoPii(authErr.errorCode + ":" + authErr.errorMessage);
+        reject(authErr);
+        return null;
+      }
+      // else proceed with login
+      else {
+        if (userContainedClaims) {
+          this.logger.verbose("Skipped cache lookup since claims were given.");
+        } else {
+          this.logger.verbose("Token is not in cache for scope:" + scope);
```
Comment (jasonnutter) [inline lib/msal-core/src/UserAgentApplication.ts]: The value of the log message could be created as a ternary.

File: lib/msal-core/src/UserAgentApplication.ts (hunk 9 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: Doesn't need it, it is only used in user agent application.

File: lib/msal-core/src/UserAgentApplication.ts (hunk 13 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: Changed the name of the function argument

File: lib/msal-core/src/UserAgentApplication.ts (hunk 14 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: done

File: lib/msal-core/src/UserAgentApplication.ts (hunk 11 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: done

File: lib/msal-core/src/Utils.ts (hunk 12 above)
Comment (pkanher617) [reply lib/msal-core/src/Utils.ts]: done

File: lib/msal-core/src/UserAgentApplication.ts (hunk 10 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: done

File: lib/msal-core/src/UserAgentApplication.ts [hunk 15]
```diff
@@ -303,122 +303,45 @@ export class UserAgentApplication {
     }
   }
 
-  //#endregion
+  private successHandler(interactionType: InteractionType, response: AuthResponse, resolve?: any) : void {
```
Comment (DarylThayil) [inline lib/msal-core/src/UserAgentApplication.ts]: should this be named something else ? success and error handler is a little vague, maybe something that represents the flows?

`tokenSuccessHandler` or something ?

File: lib/msal-core/src/UserAgentApplication.ts [hunk 16]
```diff
@@ -303,122 +303,45 @@ export class UserAgentApplication {
...
+      }
+    } else if (interactionType === Constants.interactionTypePopup) {
+      resolve(response);
+    } else {
+      throw ClientAuthError.createInvalidInteractionTypeError();
+    }
+  }
 
-  //#region Redirect Flow
+  private errorHandler(interactionType: InteractionType, authErr: AuthError, response: AuthResponse, reject?: any) : void {
```
Comment (DarylThayil) [inline lib/msal-core/src/UserAgentApplication.ts]: How much of these things can be factored out into their own file? 
You have done the work to separate them, could they also be separated from state?

File: lib/msal-core/src/Constants.ts [hunk 17]
```diff
@@ -80,6 +80,9 @@ export class Constants {
 
   static get cacheLocationLocal(): CacheLocation { return "localStorage"; }
   static get cacheLocationSession(): CacheLocation { return "sessionStorage"; }
+
+  static get interactionTypeRedirect(): InteractionType { return "redirectInteraction"; }
```
Comment (DarylThayil) [inline lib/msal-core/src/Constants.ts]: Might be better to use an enum here 

```js
enum InteractionType {
    Redirect, 
    PopUp
}
```

File: lib/msal-core/src/UserAgentApplication.ts [hunk 18]
```diff
@@ -432,116 +355,70 @@ export class UserAgentApplication {
...
-    return new Promise<AuthResponse>((resolve, reject) => {
-      // Fail if login is already in progress
-      if (this.loginInProgress) {
-        return reject(ClientAuthError.createLoginInProgressError());
-      }
+  private acquireTokenInteractive(interactionType: InteractionType, isLoginCall: boolean, request?: AuthenticationParameters, resolve?: any, reject?: any): void {
+
+    // If already in progress, do not proceed
+    if (this.loginInProgress || this.acquireTokenInProgress) {
+      let thrownError = this.loginInProgress ? ClientAuthError.createLoginInProgressError() : ClientAuthError.createAcquireTokenInProgressError();
```
Comment (DarylThayil) [inline lib/msal-core/src/UserAgentApplication.ts]: const on these

File: lib/msal-core/src/UserAgentApplication.ts [hunk 19]
```diff
@@ -432,116 +355,70 @@ export class UserAgentApplication {
...
+   * To renew idToken, please pass clientId as the only scope in the Authentication Parameters
    */
-  loginPopup(request?: AuthenticationParameters): Promise<AuthResponse> {
-    // Creates navigate url; saves value in cache; redirect user to AAD
-    return new Promise<AuthResponse>((resolve, reject) => {
-      // Fail if login is already in progress
-      if (this.loginInProgress) {
-        return reject(ClientAuthError.createLoginInProgressError());
-      }
+  private acquireTokenInteractive(interactionType: InteractionType, isLoginCall: boolean, request?: AuthenticationParameters, resolve?: any, reject?: any): void {
```
Comment (DarylThayil) [inline lib/msal-core/src/UserAgentApplication.ts]: Is there a way to do this without passing a boolean to clean up api a little bit?

File: lib/msal-core/src/UserAgentApplication.ts [hunk 20]
```diff
@@ -432,116 +355,70 @@ export class UserAgentApplication {
...
+        thrownError,
+        stateOnlyResponse,
+        reject);
+      return;
+    }
 
-      // if extraScopesToConsent is passed, append them to the login request
-      let scopes: Array<string> = this.appendScopes(request);
+    // if extraScopesToConsent is passed in loginCall, append them to the login request
+    let scopes: Array<string> = isLoginCall ? this.appendScopes(request) : request.scopes;
```
Comment (DarylThayil) [inline lib/msal-core/src/UserAgentApplication.ts]: const

File: lib/msal-core/src/UserAgentApplication.ts [hunk 21]
```diff
@@ -550,200 +427,244 @@ export class UserAgentApplication {
...
-      // Cache the state, nonce, and login request data
-      this.cacheStorage.setItem(Constants.loginRequest, window.location.href, this.inCookie);
-      this.cacheStorage.setItem(Constants.loginError, "");
+    let popUpWindow: Window;
+    if (interactionType === Constants.interactionTypePopup) {
+      // Generate a popup window
+      popUpWindow = this.openWindow("about:blank", "_blank", 1, this, resolve, reject);
+      if (!popUpWindow) {
+        // We pass reject in openWindow, we reject there during an error
+        return;
```
Comment (DarylThayil) [inline lib/msal-core/src/UserAgentApplication.ts]: this seems like an error, at the least it should have a logger.error?

File: lib/msal-core/src/UserAgentApplication.ts [hunk 22]
```diff
@@ -550,200 +427,244 @@ export class UserAgentApplication {
...
+      }
+    }
 
-      this.cacheStorage.setItem(Constants.nonceIdToken, serverAuthenticationRequest.nonce, this.inCookie);
+    acquireTokenAuthority.resolveEndpointsAsync().then(() => {
+      // On Fulfillment
+      let responseType: string;
```
Comment (DarylThayil) [inline lib/msal-core/src/UserAgentApplication.ts]: const responseType = isLoginCall ?  responseType = ResponseTypes.id_token : responseType = this.getTokenType(account, scopes, false);

File: lib/msal-core/src/UserAgentApplication.ts [hunk 23]
```diff
@@ -550,200 +427,244 @@ export class UserAgentApplication {
...
         serverAuthenticationRequest = this.populateQueryParams(account, request, serverAuthenticationRequest);
+      }
+      //if user didn't pass login_hint/sid and adal's idtoken is present, extract the login_hint from the adalIdToken
+      else if (!account && !Utils.isEmpty(adalIdToken)) {
+        // if adalIdToken exists, extract the SSO info from the same
+        const adalIdTokenObject = Utils.extractIdToken(adalIdToken);
+        this.logger.verbose("ADAL's idToken exists. Extracting login information from ADAL's idToken ");
+        serverAuthenticationRequest = this.populateQueryParams(account, null, serverAuthenticationRequest, adalIdTokenObject);
+      }
+      let userContainedClaims = request.claimsRequest || serverAuthenticationRequest.claimsValue;
```
Comment (DarylThayil) [inline lib/msal-core/src/UserAgentApplication.ts]: const

File: lib/msal-core/src/UserAgentApplication.ts [hunk 24]
```diff
@@ -1414,8 +1221,8 @@ export class UserAgentApplication {
    * Returns deserialized portion of URL hash
    * @param hash
    */
-  private deserializeHash(hash: string) {
-    hash = this.getHash(hash);
+  private deserializeHash(urlFragment: string) {
+    let hash = Utils.getHashFromUrl(urlFragment);
```
Comment (DarylThayil) [inline lib/msal-core/src/UserAgentApplication.ts]: const

File: lib/msal-core/src/Utils.ts [hunk 25]
```diff
@@ -538,6 +538,23 @@ export class Utils {
     return url;
   }
 
+  /**
+   * @hidden
+   * @ignore
+   *
+   * Returns the anchor part(#) of the URL
+   */
+  static getHashFromUrl(urlStringOrFragment: string): string {
```
Comment (DarylThayil) [inline lib/msal-core/src/Utils.ts]: test this?

File: lib/msal-core/src/error/InteractionRequiredAuthError.ts [hunk 26]
```diff
@@ -27,6 +27,14 @@ export class InteractionRequiredAuthError extends ServerError {
         Object.setPrototypeOf(this, InteractionRequiredAuthError.prototype);
     }
 
+    static isInteractionRequiredError(errorString: string) : boolean {
```
Comment (DarylThayil) [inline lib/msal-core/src/error/InteractionRequiredAuthError.ts]: add test

File: lib/msal-core/src/UserAgentApplication.ts (hunk 16 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: A lot of this work is done in later PRs

File: lib/msal-core/src/UserAgentApplication.ts (hunk 15 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: responseHandler and authErrorHandler? Since those are the objects we are sending back?

File: lib/msal-core/src/UserAgentApplication.ts (hunk 21 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: we reject with an error in openWindow as it says in the comment there. I am open to adding additional logging.

File: lib/msal-core/src/Constants.ts (hunk 17 above)
Comment (pkanher617) [reply lib/msal-core/src/Constants.ts]: I will keep this consistent with cache location for now, update to enums in a future PR

File: lib/msal-core/src/UserAgentApplication.ts (hunk 19 above)
Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: There isn't really a good way to do this since we need to identify which API was called internally. I understand that the internal function signature is a little messy, it will be cleaner when we refactor our internal promise implementation.

File: lib/msal-core/src/Utils.ts [hunk 27]
```diff
@@ -538,6 +538,23 @@ export class Utils {
...
+    const hashIndex1 = urlStringOrFragment.indexOf("#");
+    const hashIndex2 = urlStringOrFragment.indexOf("#/");
+    if (hashIndex2 > -1) {
+      return urlStringOrFragment.substring(hashIndex2 + 2);
+    } else if (hashIndex1 > -1) {
```
Comment (jasonnutter) [inline lib/msal-core/src/Utils.ts]: Nit pick: Don't need the `else` since you are returning in the above `if` block.

```js
if (condition1) {
  return ...;
}

if (condition2) {
  return ...;
}

return ...;
```

File: lib/msal-core/src/UserAgentApplication.ts (hunk 16 above)
Comment (jasonnutter) [reply lib/msal-core/src/UserAgentApplication.ts]: `reject` is a function, correct?

Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: correct

File: lib/msal-core/src/Utils.ts (hunk 27 above)
Comment (pkanher617) [reply lib/msal-core/src/Utils.ts]: Done

File: lib/msal-core/src/Utils.ts (hunk 25 above)
Comment (pkanher617) [reply lib/msal-core/src/Utils.ts]: Done

File: lib/msal-core/src/error/InteractionRequiredAuthError.ts (hunk 26 above)
Comment (pkanher617) [reply lib/msal-core/src/error/InteractionRequiredAuthError.ts]: Done

File: lib/msal-core/src/UserAgentApplication.ts [hunk 28]
```diff
@@ -1168,11 +943,13 @@ export class UserAgentApplication {
...
    * @param {string} urlNavigate - URL of the authorization endpoint
    */
-  private promptUser(urlNavigate: string) {
+  private navigateWindow(urlNavigate: string, popupWindow?: Window) {
     // Navigate if valid URL
     if (urlNavigate && !Utils.isEmpty(urlNavigate)) {
-      this.logger.infoPii("Navigate to:" + urlNavigate);
-      window.location.replace(urlNavigate);
+      let navigateWindow: Window = popupWindow ? popupWindow : window;
+      let logMessage: string = popupWindow ? "Navigated Popup window to:" + urlNavigate : "Navigate to:" + urlNavigate;
```
Comment (jasonnutter) [inline lib/msal-core/src/UserAgentApplication.ts]: Rereading this noticing that `popupWindow` will never be falsey, so the second log message condition will never happen.

Comment (pkanher617) [reply lib/msal-core/src/UserAgentApplication.ts]: I'm not sure what you mean, popupWindow will be falsey if navigateWindow is called in loginRedirect or acquireTokenRedirect, since we don't pass a window object.

Comment (jasonnutter) [reply lib/msal-core/src/UserAgentApplication.ts]: Ah whoops misread it, disregard.

**Comment by sameerag** [comment]
Me and @pkanher617 synced up offline and closed most review comments/follow-ups I had. We also created more internal PRs for future work to track some changes we wanted to get in later. I want to do a quick peek for a few more changes and will approve tomorrow.

File: lib/msal-core/src/UserAgentApplication.ts (hunk 15 above)
Comment (DarylThayil) [reply lib/msal-core/src/UserAgentApplication.ts]: willing to punt, as we refactor if we think there are better names can solve then

File: lib/msal-core/src/UserAgentApplication.ts (hunk 19 above)
Comment (DarylThayil) [reply lib/msal-core/src/UserAgentApplication.ts]: sounds good
